src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1363,608 +1450,524 @@ MulticopterPositionControl::do_position_control(float dt)
 	_run_pos_control = true;
 	_run_alt_control = true;
 
-	/* select control source */
 	if (_control_mode.flag_control_manual_enabled) {
 		/* manual control */
 		control_manual(dt);
 		_mode_auto = false;
 
-	} else if (_control_mode.flag_control_offboard_enabled) {
-		/* offboard control */
-		control_offboard(dt);
-		_mode_auto = false;
-
 	} else {
-		/* AUTO */
-		control_auto(dt);
+		control_non_manual(dt);
 	}
+}
 
-	/* weather-vane mode for vtol: disable yaw control */
-	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.disable_mc_yaw_control == true) {
-		_att_sp.disable_mc_yaw_control = true;
-
-	} else {
-		/* reset in case of setpoint updates */
-		_att_sp.disable_mc_yaw_control = false;
+void
+MulticopterPositionControl::control_position(float dt)
+{
+	/* run position & altitude controllers, if enabled (otherwise use already computed velocity setpoints) */
+	if (_run_pos_control) {
+		_vel_sp(0) = (_pos_sp(0) - _pos(0)) * _params.pos_p(0);
+		_vel_sp(1) = (_pos_sp(1) - _pos(1)) * _params.pos_p(1);
 	}
 
-	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
-	    && _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_IDLE) {
-		/* idle state, don't run controller and set zero thrust */
-		_R_setpoint.identity();
+	// guard against any bad velocity values
 
+	bool velocity_valid = PX4_ISFINITE(_pos_sp_triplet.current.vx) &&
```

Comment:
we also should try to check for PX4_ISFINITE at one common place (or maybe in th estimator as you proposed)